File: LeetCode/python/sitao/2_add_two_numbers.py
# Definition for singly-linked list.
class ListNode:
	def __init__(self, x):
		self.val = x
		self.next = None

class Solution:
	# @param {ListNode} l1
	# @param {ListNode} l2
	# @return {ListNode}
	def addTwoNumbers(self, l1, l2):
		head = ListNode(0)
		answer = ListNode(0)
		head.next = answer
		answer_p = head
		p1 = l1
		p2 = l2
		carry = 0
		while ((p1 != None) or (p2 != None) or (carry != 0)):
			if answer_p.next == None:
				answer_p.next = ListNode(0)
			answer_p = answer_p.next
			if ((p1 != None) and (p2 != None)):
				digit_sum = p1.val + p2.val + carry
				p1 = p1.next
				p2 = p2.next
			elif ((p1 != None) and (p2 == None)):
				digit_sum = p1.val + carry
				p1 = p1.next
			elif ((p1 == None) and (p2 != None)):
				digit_sum = p2.val + carry
				p2 = p2.next
			else:
				digit_sum = carry
			[carry, answer_p.val] = [0, digit_sum] if digit_sum < 10 else [1, digit_sum-10]
		return answer


def print_list(lst):
	pt = lst
	print_str = str(pt.val)
	while pt.next != None:
		pt = pt.next
		print_str += '->'+ str(pt.val)
	print print_str

a = ListNode(4)
# a.next = ListNode(4)
# a.next.next = ListNode(3)

b = ListNode(5)
# b.next = ListNode(6)
# b.next.next = ListNode(4)

print_list(a)
print_list(b)

print_list(Solution().addTwoNumbers(a, b))

File: LeetCode/cpp/zhaohui/2-AddTwoNumbers.cpp
#include <iostream>
#include <vector>
#include<algorithm>
using namespace std;
//Definition for singly-linked list.
struct ListNode {
	int val;
	ListNode *next;
	ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
	ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
		ListNode* pNode1 = l1,*pNode2 = l2;
		int temp,carry=0;
		ListNode* tNode = l1;
		while(pNode1!=NULL && pNode2 != NULL){
			temp = (pNode1->val + pNode2->val + carry);
			carry = temp / 10;
			pNode1->val = temp%10;
			tNode = pNode1;
			pNode1 = pNode1->next;
			pNode2 = pNode2->next;
		}
		if(pNode1 == NULL){
			tNode->next = pNode2;
			pNode1 = pNode2;
		}
		while(pNode1!=NULL){
			temp = (pNode1->val + carry);
			carry = temp / 10;
			pNode1->val = temp%10;
			tNode = pNode1;
			pNode1 = pNode1->next;
			//pNode2 = pNode2->next;
		}
		if(carry>0){
			tNode->next = new ListNode(carry); 
		}
		return l1;
	}
	//这个方法内存不够用了
	ListNode* addTwoNumbers1(ListNode* l1, ListNode* l2) {
		int temp = l1->val + l2->val;
		int carry = temp/10;
		ListNode* resHead = new ListNode(temp % 10);
		ListNode* pNode = resHead ;
		ListNode*pNode1 = l1->next, *pNode2 = l2->next;
		while (pNode1 != NULL && pNode2 != NULL) {
			temp = (pNode1->val + pNode2->val + carry);
			carry = temp / 10;
			pNode->next = new ListNode(temp % 10);
			pNode = pNode->next;
			pNode1 = pNode1->next;
			pNode2 = pNode2->next;
		}
		while (pNode1 != NULL) {
			temp = (pNode1->val + carry);
			carry = temp / 10;
			pNode->next = new ListNode(temp % 10);
			pNode = pNode->next;
			pNode1 = pNode1->next;
		}
		while (pNode2 != NULL) {
			temp = (pNode2->val + carry);
			carry = temp / 10;
			pNode->next = new ListNode(temp % 10);
			pNode = pNode->next;
			pNode1 = pNode2->next;
		}
		if (carry>0) {
			pNode = new ListNode(carry);
		}
		return resHead;
	}
};

int main() {
	ListNode* headNode1 = new ListNode(2);
	headNode1->next = new ListNode(4);
	headNode1->next->next = new ListNode(3);
	ListNode* headNode2 = new ListNode(5);
	headNode2->next = new ListNode(6);
	headNode2->next->next = new ListNode(4);
	Solution solution;
	ListNode* resHead = solution.addTwoNumbers(headNode1, headNode2);
	ListNode* pNode = resHead;
	while (pNode != NULL) {
		cout << pNode->val << endl;
		pNode = pNode->next;
	}
	return 0;
}